src/main/haskell/kore/src/Kore/Attribute/Parser.hs

```diff
@@ -0,0 +1,280 @@
+
+  The parser can read the 'AttributeMap' through the 'MonadReader' instance.
+
+  The parser throws and catches errors with 'throwError' and 'catchError' from
+  'MonadError'.
+
+ -}
+newtype Parser a =
+    Parser
+    { getParser :: ReaderT AttributeMap (Either (Error ParseError)) a }
+  deriving (Applicative, Functor, Monad)
+
+deriving instance MonadError (Error ParseError) Parser
+
+deriving instance MonadReader AttributeMap Parser
+
+instance Alternative Parser where
+    empty = koreFail "parse error"
+    -- If both actions fail, prefer the first failure.
+    (<|>) a b = catchError a (\e -> catchError b (\_ -> throwError e))
+
+{- | Run an attribute 'Parser' with the given list of attributes.
+ -}
+runParser :: Parser a -> Attributes -> Either (Error ParseError) a
+runParser Parser { getParser } (Attributes attrs) =
+    runReaderT getParser =<< foldM insertAttr HashMap.empty attrs
+  where
+    insertAttr attrMap attr =
+        case attr of
+            KoreObjectPattern (ApplicationPattern app) ->
+                insertApplication attrMap app
+            _ -> koreFail "expected object-level application pattern"
+
+    -- Insert the application arguments into the attribute map,
+    -- on top of any argument lists already present.
+    insertApplication
+        attrMap
+        Application
+        { applicationSymbolOrAlias
+        , applicationChildren = args
+        }
+      =
+        let
+            SymbolOrAlias { symbolOrAliasConstructor } = applicationSymbolOrAlias
+            Id { getId = attrId } = symbolOrAliasConstructor
+            insertArgs = Just . (:|) (Occurrence args) . maybe [] NonEmpty.toList
+        in
+            pure (HashMap.alter insertArgs attrId attrMap)
+
+{- | Wrap the parser in a context for the named attribute.
+
+ -}
+withContext
+    :: String  -- ^ attribute name
+    -> Parser a  -- ^ attribute parser
+    -> Parser a
+withContext attr = Kore.Error.withContext ("attribute '" ++ attr ++ "'")
+
+{- | Collect the argument lists of each occurrence of the attribute.
+
+  @parseAttribute@ signals failure if the attribute is not present or returns a
+  'NonEmpty' list containing the argument lists at each occurrence of the
+  attribute.
+
+ -}
+parseAttribute
+    :: String  -- ^ attribute name
+    -> Parser (NonEmpty Occurrence)
+parseAttribute key =
+    do
+        attrMap <- Reader.ask
+        case HashMap.lookup key attrMap of
+            Nothing -> koreFail ("no attribute found matching: " ++ key)
+            Just occurs -> pure occurs
+
+{- | Parse a key-only attribute.
+
+  A key-only attribute has no arguments. @parseKeyAttribute@ signals failure if
+  the attribute is not present or if it is present with the wrong number of
+  arguments.
+
+  See also: 'hasKeyAttribute'
+
+ -}
+parseKeyAttribute
+    :: String  -- ^ attribute name
+    -> Parser ()
+parseKeyAttribute key =
+    parseAttribute key >>= withContext key . (oneOccurrence >=> noArguments)
```

Comment:
I cannot answer that question myself because I wrote the code, but if you are asking, then the answer must be, "No, the original is not easier to read." :smiley: